File: LHDS.Core/Models/Foundations/DatasetSpecifications/DatasetSpecification.cs
// ---------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------

using System;
using System.Collections.Generic;
using LHDS.Core.Models.Bases;
using LHDS.Core.Models.Foundations.DataSets;
using LHDS.Core.Models.Foundations.SpecificationObjects;
using Microsoft.AspNetCore.Mvc.ModelBinding;

namespace LHDS.Core.Models.Foundations.DataSetSpecifications
{
    public class DataSetSpecification : IKey, IAudit
    {
        public Guid Id { get; set; }
        public Guid DataSetId { get; set; }
        public string SupplierSpecificationVersion { get; set; } = string.Empty;
        public string OurSpecificationVersion { get; set; } = string.Empty;
        public string Notes { get; set; } = string.Empty;
        public bool IsMultiAuthorPerBatch { get; set; }
        public string EntityChangeSynchronisation { get; set; } = string.Empty;
        public DateTimeOffset? DateReleased { get; set; }
        public DateTimeOffset? DateImplemented { get; set; }
        public DateTimeOffset? DateSuperseded { get; set; }
        public Guid? SupersededById { get; set; }
        public Guid? PresededById { get; set; }
        public bool IsPublished { get; set; }
        public bool IsActive { get; set; }
        public DateTimeOffset? ActiveFrom { get; set; }
        public DateTimeOffset? ActiveTo { get; set; }
        public string CreatedBy { get; set; } = string.Empty;
        public string UpdatedBy { get; set; } = string.Empty;
        public DateTimeOffset UpdatedDate { get; set; }
        public DateTimeOffset CreatedDate { get; set; }

        [BindNever]
        public DataSet? DataSet { get; set; } = null!;

        [BindNever]
        public List<SpecificationObject> SpecificationObjects { get; set; } = new List<SpecificationObject>();

        [BindNever]
        public List<DataSetSpecification> SupersededBy { get; set; } = new List<DataSetSpecification>();

        [BindNever]
        public List<DataSetSpecification> PresededBy { get; set; } = new List<DataSetSpecification>();
    }
}

File: LHDS.AdminPortal.Configurations/lhds.adminportal.configurations.client/src/models/dataSetSpecifications/dataSetSpecification.ts
import { DataSet } from '../dataSets/dataSet';

export class DataSetSpecification {
    public id: string;
    public dataSetId: string;
    public supplierSpecificationVersion: string;
    public ourSpecificationVersion?: string;
    public notes?: string;
    public isMultiAuthorPerBatch: boolean;
    public entityChangeSynchronisation?: string;
    public dateReleased?: Date;
    public dateImplementation?: Date;
    public dateSuperseded?: Date;
    public presededById?: Guid;
    public supersededById?: Guid;
    public isPublished: boolean;
    public isActive: boolean;
    public activeFrom?: Date;
    public activeTo?: Date;
    public createdBy?: string;
    public createdDate?: Date;
    public updatedBy?: string;
    public updatedDate?: Date;
    public dataSet?: DataSet;

    constructor(dataSetSpecification: DataSetSpecification) {
        this.id = dataSetSpecification.id ? dataSetSpecification.id :"";
        this.dataSetId = dataSetSpecification.dataSetId;
        this.supplierSpecificationVersion = dataSetSpecification.supplierSpecificationVersion;
        this.ourSpecificationVersion = dataSetSpecification.ourSpecificationVersion;
        this.notes = dataSetSpecification.notes;
        this.isMultiAuthorPerBatch = dataSetSpecification.isMultiAuthorPerBatch;
        this.entityChangeSynchronisation = dataSetSpecification.entityChangeSynchronisation;
        this.dateReleased = dataSetSpecification.dateReleased;
        this.dateImplementation = dataSetSpecification.dateImplementation;
        this.dateSuperseded = dataSetSpecification.dateSuperseded;
        this.presededById = dataSetSpecification.presededById;
        this.supersededById = dataSetSpecification.supersededById;
        this.isPublished = dataSetSpecification.isPublished;
        this.isActive = dataSetSpecification.isActive;
        this.activeFrom = dataSetSpecification.activeFrom;
        this.activeTo = dataSetSpecification.activeTo;
        this.createdBy = dataSetSpecification.createdBy;
        this.createdDate = new Date(dataSetSpecification.createdDate);
        this.updatedBy = dataSetSpecification.updatedBy;
        this.updatedDate = new Date(dataSetSpecification.updatedDate);

        if (dataSetSpecification.dataSet !== undefined && dataSetSpecification.dataSet !== null) {
            this.dataSet = new DataSet(dataSetSpecification.dataSet);
        }
    }
}